### ISSUE TITLE
[Feature]: Check for nodes cordoned by karpenter

### ISSUE THREAD

**Issue by quercus-carsten** [cloudnative-pg/cloudnative-pg#5299]
### Is there an existing issue already for this feature request/idea?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new feature request to be evaluated.

### What problem is this feature going to solve? Why should it be added?

At the moment, when a primary runs on EKS nodes managed by karpenter, you will have to manually failover if a node is replaced or removed. This interrupts the automatic provisioning/deprovisioning of nodes.

Kubernetes uses the "node.kubernetes.io/unschedulable" taint and the "spec.unschedulable" field on a node to mark it as cordoned. The operator checks nodes for this information.
If a primary database is running on a node which was cordoned the operator will promote a secondary on another node.
This allows to evict the cordoned node without blocking the process due to the PDB of the primary.

Karpenter does not rely on "node.kubernetes.io/unschedulable" taint and/or the "spec.unschedulable" field.
Next to other things, the project basically needs to know that karpenter was the one that initiated the tainting.

### Describe the solution you'd like

Karpenter taints nodes with: "karpenter.sh/disruption:NoSchedule"
https://karpenter.sh/docs/concepts/disruption/#disruption-controller

It would be nice if the operator could also check for an additional/alternative taint like this.

### Describe alternatives you've considered

You can set "enablePDB: false" - but this is officially not recommended for production environments.

### Backport?

N/A

### Are you willing to actively contribute to this feature?

No

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by markwebsterUK** [comment]
Ahhh, so _that's_ what's going on! Would be great to implement this as a custom field.

**Comment by danielloader** [comment]
I too found this after working out why my maximum age on karpenter issued notes wasn't cycling the nodes out and instead blocked by postgres instances. Thanks OP for working this out.

**Comment by c-thiel** [comment]
Ran into the same problem, thanks for working it out!
Karpenter community is also arguing about this, but they seem to prefer to stick to their current taint and not leverage the k8s ones:
https://github.com/kubernetes-sigs/karpenter/issues/1152

**Comment by mark-webster-catalyst** [comment]
Any movement on this? It's making clusters managed by Karpenter unable to reliably consolidate or roll nodes without user interaction.

**Comment by chris-gooch** [comment]
We are hitting this issue as well

**Comment by quercus-carsten** [comment]
Yep, i couldn't agree more :). This is going to be awesome - thank you!
